py/gc_long_lived.c

```diff
@@ -0,0 +1,132 @@
+/*
+ * This file is part of the MicroPython project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2018 Scott Shawcroft for Adafruit Industries LLC
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "py/emitglue.h"
+#include "py/gc_long_lived.h"
+#include "py/gc.h"
+
+mp_obj_fun_bc_t *make_fun_bc_long_lived(mp_obj_fun_bc_t *fun_bc, uint8_t max_depth) {
+    #ifndef MICROPY_ENABLE_GC
+    return fun_bc;
+    #endif
+    if (fun_bc == NULL || fun_bc == mp_const_none || max_depth == 0) {
+        return fun_bc;
+    }
+    fun_bc->bytecode = gc_make_long_lived((byte*) fun_bc->bytecode);
+    fun_bc->globals = make_dict_long_lived(fun_bc->globals, max_depth - 1);
+    for (uint32_t i = 0; i < gc_nbytes(fun_bc->const_table) / sizeof(mp_obj_t); i++) {
+        // Skip things that aren't allocated on the heap (and hence have zero bytes.)
+        if (gc_nbytes((byte *)fun_bc->const_table[i]) == 0) {
+            continue;
+        }
+        // Try to detect raw code.
+        mp_raw_code_t* raw_code = MP_OBJ_TO_PTR(fun_bc->const_table[i]);
+        if (raw_code->kind == MP_CODE_BYTECODE) {
+            raw_code->data.u_byte.bytecode = gc_make_long_lived((byte*) raw_code->data.u_byte.bytecode);
+            // TODO(tannewt): Do we actually want to recurse here?
+            raw_code->data.u_byte.const_table = gc_make_long_lived((byte*) raw_code->data.u_byte.const_table);
+        }
+        ((mp_uint_t *) fun_bc->const_table)[i] = (mp_uint_t) make_obj_long_lived(
+            (mp_obj_t) fun_bc->const_table[i], max_depth - 1);
+
+    }
+    fun_bc->const_table = gc_make_long_lived((mp_uint_t*) fun_bc->const_table);
+    // extra_args stores keyword only argument default values.
+    size_t words = gc_nbytes(fun_bc) / sizeof(mp_uint_t*);
+    for (size_t i = 0; i < words - 4; i++) {
```

Comment:
What's the `4`? Is that number of bytes? could it be 8 on 64-bit-word impls?

Reply:
Its the number of pointers stored in mp_obj_fun_bc_t before the extra_args array. Is there another way to get the array length?

Reply:
The struct defn is:
```
typedef struct _mp_obj_fun_bc_t {
    mp_obj_base_t base;
    mp_obj_dict_t *globals;         // the context within which this function was defined
    const byte *bytecode;           // bytecode for the function
    const mp_uint_t *const_table;   // constant table
    // the following extra_args array is allocated space to take (in order):
    //  - values of positional default args (if any)
    //  - a single slot for default kw args dict (if it has them)
    //  - a single slot for var args tuple (if it takes them)
    //  - a single slot for kw args dict (if it takes them)
    mp_obj_t extra_args[];
} mp_obj_fun_bc_t;
```
I'm not sure why it doesn't say `[4]`. then I think you could use `sizeof()`. And if it's a VLA (variable length array), you can use `sizeof()` also. Found this: https://stackoverflow.com/questions/14995870/behavior-of-sizeof-on-variable-length-arrays-c-only.

But I'm not sure this is worth fixing.

Reply:
I guess it's allocated separately and assigned to there?

Reply:
Its done through a cast so I'm not sure if sizeof would work: https://github.com/adafruit/circuitpython/blob/master/py/objfun.c#L356